From d068fb0caae8c8db97ee59ba400929f5f97f8be1 Mon Sep 17 00:00:00 2001
Message-Id: <d068fb0caae8c8db97ee59ba400929f5f97f8be1.1650123861.git.github@sicherha.de>
From: Christoph Erhardt <github@sicherha.de>
Date: Tue, 22 Feb 2022 21:21:18 +0100
Subject: [PATCH] Fix compatibility with libstdc++ < 9

mimalloc does not override the `nothrow` variants of the `delete`
operator because it assumes that their implementation in libstdc++
redirects to the default `delete` operators. This is not the case for
libstdc++ < 9, where `std::free()` is called directly.

Upstream bug: https://gcc.gnu.org/bugzilla/show_bug.cgi?id=68210

This patch ensures that the `nothrow` `delete` operators are overridden
as well.
---
 main.cc | 7 +++++++
 1 file changed, 7 insertions(+)

diff --git a/main.cc b/main.cc
index b390b0f7..5c1f5450 100644
--- a/main.cc
+++ b/main.cc
@@ -6,6 +6,13 @@
 
 #ifdef USE_SYSTEM_MIMALLOC
 #include <mimalloc-new-delete.h>
+
+// Additionally override the nothrow delete operators for compatibility with
+// buggy libstdc++ < 9
+#if MI_MALLOC_VERSION < 206
+void operator delete(void* p, const std::nothrow_t&) noexcept   { mi_free(p); }
+void operator delete[](void* p, const std::nothrow_t&) noexcept { mi_free(p); }
+#endif
 #endif
 
 namespace mold {
-- 
2.35.1

